Assertj (#175)
OICR/softeng-blog: _posts/dusan_andric/2019-08-19-assertJ.md

---
layout: post
title:  "Don't use AssertJ"
breadcrumb: true
author: dusan_andric
date: 2019-08-19
categories: dusan_andric
tags:
    - Java
    - JDK11
    - Test
    - JUnit
    - AssertJ
teaser:
    info: When practicality trumps ergonomics. 
    image: dusan_andric/assertj/assertj.png
header: 
    version: small
    title: Software Engineering Blog
    image: header-logo-crop.png
    icon: icon-blog
---

## Introduction

<figure>
    <img src="{{site.urlimg}}dusan_andric/assertj/assertj.png" />
    <figcaption>An acceptable amount of compile time.</figcaption>
</figure>

Dear reader, look at the screenshot in the title image. Now imagine instead of the test-compile phase taking the pictured amount of time, it instead took ~5 minutes. Now imagine solving this issue by simply swapping what assertion library was being used in your unit tests.

## Wait really?
Yes.

<figure>
    <img src="{{site.urlimg}}dusan_andric/assertj/assertJ_pr.png" />
    <figcaption>The first attempt at fixing compilation times.</figcaption>
</figure>

## But why?

> Being completely honest, we do not have a have the definitive answer to what was causing this, however, I do have a hunch.

Reviewer: "we do not have a have the" => "we do not have the def..."

> We know that this became a big problem for us when we switched to Java 11. We also know that it didn't matter if we were using Oracle JDK or OpenJDK as they are within margin of error of one another. One of our team members, [Kevin](http://softeng.oicr.on.ca/blog/category/kevin_hartmann), discovered previously reported issues in JDK8 about a regression in compliation times, as well as him discovering other various issues and creating bug reports for `javac`.

Reviewer: "within margin" => "within a margin.."

> Given this lack of polish with `javac` in Java 11 and knowing there were previous type inference issues that were probably exaserbated with the addition of the `var` keyword in Java 10, I figured that type inference had to be related to the issue, or at the very least the type system itself.

Reviewer 2: ```suggestion
Given this lack of polish with `javac` in Java 11 and knowing there were previous type inference issues that were probably exacerbated with the addition of the `var` keyword in Java 10, I figured that type inference had to be related to the issue, or at the very least the type system itself.
```

> A big reason for us using AssertJ, in addition to handling exception better than JUnit4, is the fluent API it had made it a lot more ergonomic to use for developers. Code looked better, was easier to understand, and it made reuse easier.

Reviewer 2: ```suggestion
A big reason for us using AssertJ, in addition to handling exception better than JUnit4, is the fluent API that had made it a lot more ergonomic to use for developers. Code looked better, was easier to understand, and it made reuse easier.
```

If you are unfamiliar with the API of AssertJ, I have an example here:
~~~java
    assertThat(myList).hasSize(n).containsAnyOf(matchObj)
~~~

You can imagine that `assertThat` not only being a heavily overloaded method and making use of generics, also allows you to branch out to many different similarly generic, heavily overloaded methods that are context sensitive on the return type of the previous callable.

> Another fact to consider is that `javac` is not sitting idle during this long compilation time, its going full tilt on my machine maxing out the cores it is running on.

Reviewer: its going
it's going

> My hypothesis is that at some stage in `javac`, either type inference, or during the type erasure where it is bounding the generic types, that the root of the problem lies. It must be some computation of high complexity, maybe exponential based on the hight of the type hierarchy for AssertJ.

Reviewer: hight  => height?

> My hypothesis is that at some stage in `javac`, either type inference, or during the type erasure where it is bounding the generic types, that the root of the problem lies. It must be some computation of high complexity, maybe exponential based on the hight of the type hierarchy for AssertJ.

Reviewer 2: `hight of the type hierarchy`

what's hight?

## Do I have this problem too?

> One of the symptoms we noticed is that opening a project and editing a file in IntelliJ Idea, the IDE would lint/inspect the file at a snails pace. For example if I removed an important import and hit `ctrl+s`, for several seconds the IDE would slowly make its way through the file and I could see the red ticks on the scroll bar crawl it's way down.

Reviewer 3: `at a snails pace` => `at a snail's pace`

> One of the symptoms we noticed is that opening a project and editing a file in IntelliJ Idea, the IDE would lint/inspect the file at a snails pace. For example if I removed an important import and hit `ctrl+s`, for several seconds the IDE would slowly make its way through the file and I could see the red ticks on the scroll bar crawl it's way down.

Reviewer 2: ```suggestion
One of the symptoms we noticed is that when opening a project and editing a file in IntelliJ Idea, the IDE would lint/inspect the file at a snail's pace. For example if I removed an important import and hit `ctrl+s`, for several seconds the IDE would slowly make its way through the file and I could see the red ticks on the scroll bar crawl its way down.
```

> If this truly is related to complex type hierarchies and generics I would love it other people reached out and gave us examples outside of AssertJ if for nothing more than building an understanding of what the root of the problem really is. I'd be just as happy if you were to debunk my hypothesis.

Reviewer: i would love it other people
i would love it if other people

> If this truly is related to complex type hierarchies and generics I would love it other people reached out and gave us examples outside of AssertJ if for nothing more than building an understanding of what the root of the problem really is. I'd be just as happy if you were to debunk my hypothesis.

Reviewer 3: `I would love it other people reached out` -> `I would love it IF other people reached out`

> If this truly is related to complex type hierarchies and generics I would love it other people reached out and gave us examples outside of AssertJ if for nothing more than building an understanding of what the root of the problem really is. I'd be just as happy if you were to debunk my hypothesis.

Reviewer 2: ```suggestion
If this truly is related to complex type hierarchies and generics I would love it if other people reached out and gave us examples outside of AssertJ, if for nothing more than building an understanding of what the root of the problem really is. I'd be just as happy if you were to debunk my hypothesis.
```
